File: frontend/types/anamnese.ts
/**
 * Types e Interfaces para o Sistema de Anamnese
 * Usados em AnamneseFirstConsultation e AnamneseFollowUp
 */

/**
 * Template type: Tipo de anamnese
 */
export type TemplateType = 'first_consultation' | 'follow_up';

/**
 * Status da anamnese
 */
export type AnamneseStatus = 'draft' | 'completed' | 'signed';

/**
 * Localização da dor
 */
export type PainLocation = 'Ombro' | 'Cotovelo' | 'Punho' | 'Coluna' | 'Quadril' | 'Joelho' | 'Tornozelo';

/**
 * Nível de consumo de álcool
 */
export type AlcoholConsumption = 'none' | 'occasional' | 'regular';

/**
 * Qualidade do sono
 */
export type SleepQuality = 'good' | 'fair' | 'poor';

/**
 * Nível de adesão ao tratamento
 */
export type TreatmentAdherence = 'excellent' | 'good' | 'fair' | 'poor';

/**
 * Medicação selecionada
 */
export interface Medicacao {
  id: string;
  nome: string;
  dosagem: string;
  via: 'oral' | 'injetável' | 'tópica' | 'intravenosa';
  fabricante: string;
  posologia?: string;
  duracao?: string;
  interacoes?: Array<{
    medicacao: string;
    severidade: 'leve' | 'moderada' | 'grave';
    descricao: string;
  }>;
  temContraindicacao?: boolean;
}

/**
 * Dados da Anamnese de Primeira Consulta
 */
export interface AnamneseFirstConsultationData {
  template_type: 'first_consultation';
  status: AnamneseStatus;

  // TAB 1: Queixa Principal
  chief_complaint: string;
  pain_location: string;
  symptom_duration: string;
  pain_intensity: number; // 0-10
  aggravating_factors: string;
  relieving_factors: string;

  // TAB 2: Histórico
  previous_traumas: string;
  previous_surgeries: string;
  previous_physiotherapy: string;
  previous_treatments: string;
  current_medications: Medicacao[];

  // TAB 3: Hábitos e Risco
  profession: string;
  physical_activities: string;
  sedentarism: number; // 1-10
  smoking: boolean;
  alcohol_consumption: AlcoholConsumption;
  sleep_quality: SleepQuality;

  // TAB 4: Exame Físico
  range_of_motion: string;
  specific_tests: string;
  inflammation: boolean;
  deformities: string;

  // Metadados
  created_at?: string;
}

/**
 * Dados da Anamnese de Retorno
 */
export interface AnamneseFollowUpData {
  template_type: 'follow_up';
  status: AnamneseStatus;

  last_consultation_date?: string;
  pain_intensity_last: number; // 0-10
  pain_intensity_current: number; // 0-10
  evolution_summary: string;
  new_symptoms: string;
  treatment_adherence: TreatmentAdherence;
  treatment_response: string;
  current_limitations: string;
  next_steps: string;
  additional_observations: string;

  // Metadados
  created_at?: string;
}

/**
 * Type union de dados de anamnese
 */
export type AnamneseData = AnamneseFirstConsultationData | AnamneseFollowUpData;

/**
 * Resposta da API ao criar/atualizar anamnese
 */
export interface AnamneseResponse {
  id: number;
  patient_id: number;
  template_type: TemplateType;
  status: AnamneseStatus;
  dados: AnamneseData;
  created_at: string;
  updated_at: string;
  created_by?: number;
}

/**
 * Parâmetros para criar anamnese
 */
export interface CreateAnamneseParams {
  patient_id: number;
  template_type: TemplateType;
  status: AnamneseStatus;
  dados: AnamneseData;
}

/**
 * Props do componente AnamneseTemplateSelector
 */
export interface AnamneseTemplateSelectorProps {
  patientName: string;
  onSelect: (templateType: TemplateType) => void;
}

/**
 * Props do componente AnamneseFirstConsultation
 */
export interface AnamneseFirstConsultationProps {
  patientId: number;
  patientName: string;
  onSave?: (data: AnamneseFirstConsultationData) => void;
}

/**
 * Props do componente AnamneseFollowUp
 */
export interface AnamneseFollowUpProps {
  patientId: number;
  patientName: string;
  lastConsultationDate?: string;
  lastPainIntensity?: number;
  onSave?: (data: AnamneseFollowUpData) => void;
}

/**
 * Histórico de anamnese de um paciente
 */
export interface PatientAnamneseHistory {
  id: number;
  patient_id: number;
  template_type: TemplateType;
  status: AnamneseStatus;
  created_at: string;
  created_by?: number;
  dados: AnamneseData;
}

/**
 * Metadados de anamnese para listagem
 */
export interface AnamneseListItem {
  id: number;
  patient_id: number;
  template_type: TemplateType;
  status: AnamneseStatus;
  created_at: string;
  chief_complaint?: string; // Para first_consultation
  pain_intensity_last?: number; // Para follow_up
  pain_intensity_current?: number; // Para follow_up
}

/**
 * Estatísticas de anamnese
 */
export interface AnamneseStats {
  total: number;
  first_consultations: number;
  follow_ups: number;
  completed: number;
  drafts: number;
  signed: number;
  last_anamnese_date?: string;
}

/**
 * Pain colors mapping
 */
export const PAIN_COLORS = [
  'bg-green-500',
  'bg-green-400',
  'bg-lime-400',
  'bg-yellow-300',
  'bg-yellow-400',
  'bg-orange-300',
  'bg-orange-400',
  'bg-orange-500',
  'bg-red-400',
  'bg-red-500',
  'bg-red-600',
] as const;

/**
 * Pain locations constants
 */
export const PAIN_LOCATIONS = [
  'Ombro',
  'Cotovelo',
  'Punho',
  'Coluna',
  'Quadril',
  'Joelho',
  'Tornozelo',
] as const;

/**
 * Utilidades para conversão de tipos
 */
export namespace AnamneseUtils {
  /**
   * Calcula melhora/piora da dor em porcentagem
   */
  export function calculatePainImprovement(
    painBefore: number,
    painAfter: number
  ): number {
    if (painBefore === 0) return 0;
    return Math.round(((painBefore - painAfter) / painBefore) * 100);
  }

  /**
   * Converte sleep_quality para label
   */
  export function getSleepQualityLabel(quality: SleepQuality): string {
    const labels: Record<SleepQuality, string> = {
      good: 'Boa',
      fair: 'Regular',
      poor: 'Ruim',
    };
    return labels[quality];
  }

  /**
   * Converte alcohol_consumption para label
   */
  export function getAlcoholConsumptionLabel(consumption: AlcoholConsumption): string {
    const labels: Record<AlcoholConsumption, string> = {
      none: 'Não',
      occasional: 'Ocasional',
      regular: 'Regular',
    };
    return labels[consumption];
  }

  /**
   * Converte treatment_adherence para label
   */
  export function getTreatmentAdherenceLabel(adherence: TreatmentAdherence): string {
    const labels: Record<TreatmentAdherence, string> = {
      excellent: 'Excelente (100%)',
      good: 'Boa (75%)',
      fair: 'Regular (50%)',
      poor: 'Fraca (<50%)',
    };
    return labels[adherence];
  }

  /**
   * Calcula percentual de preenchimento
   */
  export function calculateCompletionPercentage(data: AnamneseData): number {
    const keys = Object.keys(data).filter(
      (k) => !['template_type', 'status', 'created_at'].includes(k)
    );

    const filled = keys.filter((k) => {
      const value = (data as any)[k];
      if (typeof value === 'string') return value.trim().length > 0;
      if (typeof value === 'number') return true;
      if (typeof value === 'boolean') return true;
      if (Array.isArray(value)) return value.length > 0;
      return false;
    });

    return Math.round((filled.length / keys.length) * 100);
  }

  /**
   * Valida se anamnese está completa
   */
  export function isComplete(data: AnamneseData): boolean {
    if (data.template_type === 'first_consultation') {
      return (data as AnamneseFirstConsultationData).chief_complaint.trim().length > 0;
    } else {
      return (data as AnamneseFollowUpData).evolution_summary.trim().length > 0;
    }
  }
}

File: backend/app/schemas/anamnese.py
"""
Schemas Pydantic para validação de dados de Anamnese
"""

from pydantic import BaseModel, Field, validator
from typing import Optional, List, Dict, Any
from datetime import datetime
from enum import Enum


# ============ ENUMS ============

class TemplateType(str, Enum):
    FIRST_CONSULTATION = "first_consultation"
    FOLLOW_UP = "follow_up"


class AnamneseStatus(str, Enum):
    DRAFT = "draft"
    COMPLETED = "completed"
    SIGNED = "signed"


class AlcoholConsumption(str, Enum):
    NONE = "none"
    OCCASIONAL = "occasional"
    REGULAR = "regular"


class SleepQuality(str, Enum):
    GOOD = "good"
    FAIR = "fair"
    POOR = "poor"


class TreatmentAdherence(str, Enum):
    EXCELLENT = "excellent"
    GOOD = "good"
    FAIR = "fair"
    POOR = "poor"


# ============ MEDICAÇÃO ============

class MedicacaoBase(BaseModel):
    id: str
    nome: str
    dosagem: str
    via: str
    fabricante: str
    posologia: Optional[str] = None
    duracao: Optional[str] = None


class MedicacaoResponse(MedicacaoBase):
    interacoes: Optional[List[Dict[str, Any]]] = []
    temContraindicacao: Optional[bool] = False

    class Config:
        json_schema_extra = {
            "example": {
                "id": "med_001",
                "nome": "Dipirona",
                "dosagem": "500mg",
                "via": "oral",
                "fabricante": "Genérico",
                "posologia": "1 comprimido a cada 8h",
                "duracao": "7 dias",
                "interacoes": [],
                "temContraindicacao": False,
            }
        }


# ============ PRIMEIRA CONSULTA ============

class AnamneseFirstConsultationDataBase(BaseModel):
    # TAB 1: Queixa Principal
    chief_complaint: str = Field(..., min_length=1, description="Motivo da consulta")
    pain_location: str = Field(
        default="", description="Localização da dor (Ombro, Cotovelo, etc)"
    )
    symptom_duration: str = Field(default="", description="Há quanto tempo (ex: 2 semanas)")
    pain_intensity: int = Field(default=5, ge=0, le=10, description="Intensidade de 0-10")
    aggravating_factors: str = Field(default="", description="O que piora")
    relieving_factors: str = Field(default="", description="O que melhora")

    # TAB 2: Histórico
    previous_traumas: str = Field(default="", description="Traumas anteriores")
    previous_surgeries: str = Field(default="", description="Cirurgias anteriores")
    previous_physiotherapy: str = Field(
        default="", description="Fisioterapia anterior"
    )
    previous_treatments: str = Field(default="", description="Outros tratamentos")
    current_medications: List[MedicacaoResponse] = Field(
        default=[], description="Medicações atuais"
    )

    # TAB 3: Hábitos e Risco
    profession: str = Field(default="", description="Profissão")
    physical_activities: str = Field(default="", description="Atividades físicas")
    sedentarism: int = Field(default=5, ge=1, le=10, description="Nível de sedentarismo")
    smoking: bool = Field(default=False, description="É fumante?")
    alcohol_consumption: AlcoholConsumption = Field(
        default="none", description="Consumo de álcool"
    )
    sleep_quality: SleepQuality = Field(default="good", description="Qualidade do sono")

    # TAB 4: Exame Físico
    range_of_motion: str = Field(default="", description="Amplitude de movimento")
    specific_tests: str = Field(default="", description="Testes específicos realizados")
    inflammation: bool = Field(default=False, description="Presença de inflamação")
    deformities: str = Field(default="", description="Deformidades/assimetrias")

    class Config:
        json_schema_extra = {
            "example": {
                "chief_complaint": "Dor no ombro direito ao movimentar",
                "pain_location": "Ombro",
                "symptom_duration": "2 semanas",
                "pain_intensity": 7,
                "aggravating_factors": "Levantamento de peso, abdução",
                "relieving_factors": "Repouso, gelo",
                "previous_traumas": "",
                "previous_surgeries": "",
                "previous_physiotherapy": "",
                "previous_treatments": "",
                "current_medications": [],
                "profession": "Vendedor",
                "physical_activities": "Musculação",
                "sedentarism": 3,
                "smoking": False,
                "alcohol_consumption": "occasional",
                "sleep_quality": "good",
                "range_of_motion": "Limitada em abdução",
                "specific_tests": "Teste de Neer positivo",
                "inflammation": True,
                "deformities": "Sem deformidades",
            }
        }


class AnamneseFirstConsultationCreate(AnamneseFirstConsultationDataBase):
    pass


# ============ RETORNO ============

class AnamneseFollowUpDataBase(BaseModel):
    last_consultation_date: Optional[str] = None
    pain_intensity_last: int = Field(ge=0, le=10, description="Dor na última consulta")
    pain_intensity_current: int = Field(ge=0, le=10, description="Dor atual")
    evolution_summary: str = Field(..., min_length=1, description="Evolução do paciente")
    new_symptoms: str = Field(default="", description="Novos sintomas")
    treatment_adherence: TreatmentAdherence = Field(
        default="good", description="Adesão ao tratamento"
    )
    treatment_response: str = Field(default="", description="Resposta ao tratamento")
    current_limitations: str = Field(default="", description="Limitações atuais")
    next_steps: str = Field(default="", description="Próximos passos")
    additional_observations: str = Field(
        default="", description="Observações adicionais"
    )

    class Config:
        json_schema_extra = {
            "example": {
                "last_consultation_date": "2025-01-15",
                "pain_intensity_last": 8,
                "pain_intensity_current": 4,
                "evolution_summary": "Paciente evoluiu bem com o tratamento. Dor reduzida significativamente.",
                "new_symptoms": "Nenhum",
                "treatment_adherence": "good",
                "treatment_response": "Respondeu bem ao repouso e gelo",
                "current_limitations": "Ainda com dificuldade em levantar peso",
                "next_steps": "Continuar fisioterapia 2x por semana",
                "additional_observations": "Paciente aderente e motivado",
            }
        }


class AnamneseFollowUpCreate(AnamneseFollowUpDataBase):
    pass


# ============ CREATE/UPDATE ============

class AnamneseCreate(BaseModel):
    patient_id: int = Field(..., description="ID do paciente")
    template_type: TemplateType = Field(..., description="Tipo de anamnese")
    status: AnamneseStatus = Field(default="draft", description="Status da anamnese")
    dados: Dict[str, Any] = Field(..., description="Dados estruturados da anamnese")

    @validator("dados")
    def validate_dados(cls, v, values):
        """Validar estrutura de dados de acordo com template_type"""
        if "template_type" in values:
            template_type = values["template_type"]
            required_fields = (
                [
                    "chief_complaint",
                    "pain_intensity",
                    "sleep_quality",
                    "alcohol_consumption",
                ]
                if template_type == "first_consultation"
                else ["evolution_summary", "pain_intensity_current", "pain_intensity_last"]
            )

            missing = [f for f in required_fields if f not in v]
            if missing:
                raise ValueError(f"Campos obrigatórios ausentes: {missing}")

        return v


class AnamneseUpdate(BaseModel):
    status: Optional[AnamneseStatus] = None
    dados: Optional[Dict[str, Any]] = None


# ============ RESPONSES ============

class AnamneseListResponse(BaseModel):
    id: int
    patient_id: int
    template_type: TemplateType
    status: AnamneseStatus
    created_at: datetime

    class Config:
        from_attributes = True


class AnamneseResponse(BaseModel):
    id: int
    patient_id: int
    template_type: TemplateType
    status: AnamneseStatus
    dados: Dict[str, Any]
    created_at: datetime
    updated_at: datetime
    created_by: Optional[int] = None

    class Config:
        from_attributes = True
        json_schema_extra = {
            "example": {
                "id": 1,
                "patient_id": 123,
                "template_type": "first_consultation",
                "status": "completed",
                "dados": {
                    "chief_complaint": "Dor no ombro",
                    "pain_intensity": 7,
                },
                "created_at": "2025-01-20T10:30:00",
                "updated_at": "2025-01-20T10:30:00",
                "created_by": 456,
            }
        }


# ============ STATS ============

class AnamneseStats(BaseModel):
    total: int
    first_consultations: int
    follow_ups: int
    completed: int
    drafts: int
    signed: int
    last_anamnese_date: Optional[str] = None

    class Config:
        json_schema_extra = {
            "example": {
                "total": 5,
                "first_consultations": 1,
                "follow_ups": 4,
                "completed": 4,
                "drafts": 1,
                "signed": 0,
                "last_anamnese_date": "2025-01-20T10:30:00",
            }
        }
